tests/robustness/client/watch.go

```diff
@@ -20,28 +20,34 @@ import (
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
 
-	"go.etcd.io/etcd/tests/v3/framework/e2e"
 	"go.etcd.io/etcd/tests/v3/robustness/identity"
 	"go.etcd.io/etcd/tests/v3/robustness/report"
 )
 
-func CollectClusterWatchEvents(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessCluster, maxRevisionChan <-chan int64, cfg WatchConfig, baseTime time.Time, ids identity.Provider) []report.ClientReport {
+func CollectClusterWatchEvents(ctx context.Context, lg *zap.Logger, etcdEndpoints [][]string, maxRevisionChan <-chan int64, cfg WatchConfig, baseTime time.Time, ids identity.Provider) ([]report.ClientReport, bool, error) {
 	mux := sync.Mutex{}
 	var wg sync.WaitGroup
-	reports := make([]report.ClientReport, len(clus.Procs))
-	memberMaxRevisionChans := make([]chan int64, len(clus.Procs))
-	for i, member := range clus.Procs {
-		c, err := NewRecordingClient(member.EndpointsGRPC(), ids, baseTime)
-		require.NoError(t, err)
+	var everFailed bool
+	reports := make([]report.ClientReport, len(etcdEndpoints))
+	memberMaxRevisionChans := make([]chan int64, len(etcdEndpoints))
+	for i, endpoint := range etcdEndpoints {
+		c, err := NewRecordingClient(endpoint, ids, baseTime)
+		if err != nil {
+			return nil, true, err
+		}
 		memberMaxRevisionChan := make(chan int64, 1)
 		memberMaxRevisionChans[i] = memberMaxRevisionChan
 		wg.Add(1)
 		go func(i int, c *RecordingClient) {
 			defer wg.Done()
 			defer c.Close()
-			watchUntilRevision(ctx, t, c, memberMaxRevisionChan, cfg)
+
+			watchFailed := watchUntilRevision(ctx, lg, c, memberMaxRevisionChan, cfg)
+			if watchFailed && !everFailed {
+				everFailed = true
+			}
```

Comment:
This code is executed in separate goroutine, it needs to be synchronized via mutex, same like collecting reports.